### PR TITLE
Fixed CSRF handling for invalid content types

### PR DESCRIPTION
**Pull request by jroper**

### CODE REVIEW

**Comment by gmethvin** [comment]
It would be nice to clarify the documentation of `RequestHeader#contentType` to clarify what happens when a `Content-Type` header is invalid, and also for `CSRFConfig#checkContentType` to clarify that it's receiving the value of the *parsed* `Content-Type` header, _without parameters_.

I understand the motivation for breaking the API here, but I wonder if this change defeats the purpose of the `checkContentType` function in the first place. It's not clear what the difference in functionality is now between `checkContentType` and `shouldProtect`. Both functions appear to take a request header and return a boolean value indicating whether the request should be checked.

If we're going to change the API, it seems to make more sense to either remove the `checkContentType` function entirely and put all its logic into `shouldProtect`, or to add another boolean config parameter `checkInvalidContentTypes` (defaulting to true) and keep the type of `checkContentType` the same. But I'll admit I didn't take a look at the actual logic inside the filter, so there could be some special cases I'm not considering fully.

File: project/BuildSettings.scala [hunk 1]
```diff
@@ -268,6 +268,20 @@ object BuildSettings {
...
+      ProblemFilters.exclude[IncompatibleSignatureProblem]("play.filters.csrf.CSRFConfig.unapply"),
+      ProblemFilters.exclude[IncompatibleSignatureProblem]("play.filters.csrf.CSRFConfig.<init>$default$10"),
+      ProblemFilters.exclude[IncompatibleSignatureProblem]("play.filters.csrf.CSRFConfig.unapply"),
+      ProblemFilters.exclude[IncompatibleSignatureProblem]("play.filters.csrf.CSRFConfig.apply$default$10"),
+      ProblemFilters.exclude[IncompatibleSignatureProblem]("play.filters.csrf.CSRFConfig.apply"),
+      ProblemFilters.exclude[IncompatibleSignatureProblem]("play.filters.csrf.CSRFConfig.checkContentType"),
+      ProblemFilters.exclude[IncompatibleSignatureProblem]("play.filters.csrf.CSRFConfig.withContentTypes"),
+      ProblemFilters.exclude[IncompatibleSignatureProblem]("play.filters.csrf.CSRFConfig.copy"),
+      ProblemFilters.exclude[IncompatibleSignatureProblem]("play.filters.csrf.CSRFConfig.copy$default$10"),
+      ProblemFilters.exclude[IncompatibleSignatureProblem]("play.filters.csrf.CSRFConfig.this"),
```
Comment (ignasi35) [inline project/BuildSettings.scala:284]: Not a big deal, but some filters are duplicate:

```
<init>$default$10
apply$default$10
apply
unapply
```

**Comment by jroper** [comment]
I think I've just realised that we can do this without changing the API - I had thought we needed to change the API because the invalid content type check should only be done when blacklisting content types - when whitelisting, an invalid content type would fail anyway because it wouldn't be in the whitelist, and when using no lists, everything needs to be checked anyway. But then - we may as well add a condition where we always check if the content type is invalid regardless of black or white list configuration, since the result will be identical.

**Comment by ignasi35** [comment]
@Mergifyio backport 2.8.x

**Comment by ignasi35** [comment]
@Mergifyio backport 2.7.x

**Comment by ikkisoft** [comment]
Hey @gmethvin long time no talk. Thanks for helping us on this issue.

**Comment by gmethvin** [comment]
Of course @ikkisoft! Hope you're well.

**Pull request merged**